File: examples/python/webhook_verification.py
"""
Webhook Verification Example (Python)

Demonstrates HMAC-SHA256 signature verification, timestamp validation,
and idempotency checking for IntegrateWise OS webhooks.
"""

import os
import hmac
import hashlib
import json
import time
from datetime import datetime, timedelta
from flask import Flask, request, jsonify
from typing import Optional, Dict, Any

app = Flask(__name__)
WEBHOOK_SECRET = os.environ.get('WEBHOOK_SECRET', 'your-webhook-secret')
TIMESTAMP_TOLERANCE = 300  # 5 minutes in seconds
REPLAY_WINDOW = 300  # 5 minutes in seconds

# In-memory idempotency store (use Redis/DB in production)
idempotency_store: Dict[str, Dict[str, Any]] = {}


def is_timestamp_valid(timestamp: int, tolerance_seconds: int = TIMESTAMP_TOLERANCE) -> bool:
    """Verify timestamp is within tolerance window."""
    now = int(time.time())
    diff = abs(now - timestamp)
    return diff <= tolerance_seconds


def calculate_signature(secret: str, timestamp: str, body: Dict[str, Any]) -> str:
    """Calculate HMAC-SHA256 signature."""
    payload = f"{timestamp}.{json.dumps(body, sort_keys=True)}"
    hmac_obj = hmac.new(
        secret.encode('utf-8'),
        payload.encode('utf-8'),
        hashlib.sha256
    )
    return f"sha256={hmac_obj.hexdigest()}"


def verify_signature(secret: str, timestamp: str, body: Dict[str, Any], signature: str) -> bool:
    """Verify signature using timing-safe comparison."""
    expected = calculate_signature(secret, timestamp, body)
    return hmac.compare_digest(signature, expected)


def check_idempotency(idempotency_key: str) -> Optional[Dict[str, Any]]:
    """Check if idempotency key was already processed."""
    stored = idempotency_store.get(idempotency_key)
    if not stored:
        return None

    # Check if within replay window
    now = time.time() * 1000  # milliseconds
    age = now - stored['timestamp']
    if age > REPLAY_WINDOW * 1000:
        del idempotency_store[idempotency_key]
        return None

    return stored['response']


def store_idempotency(idempotency_key: str, response: Dict[str, Any]) -> None:
    """Store idempotency key with response."""
    idempotency_store[idempotency_key] = {
        'timestamp': time.time() * 1000,  # milliseconds
        'response': response
    }


async def process_webhook(connector: str, payload: Dict[str, Any]) -> Dict[str, Any]:
    """Process webhook payload."""
    # Simulate processing
    print(f"Processing webhook from {connector}: {payload}")
    
    # In real implementation:
    # 1. Parse payload
    # 2. Map to Spine entity
    # 3. Validate constraints
    # 4. Store in Spine
    # 5. Trigger downstream processing
    
    return {
        'success': True,
        'processedAt': datetime.utcnow().isoformat() + 'Z'
    }


@app.route('/webhooks/<connector>', methods=['POST'])
async def webhook_handler(connector: str):
    """Webhook endpoint handler."""
    signature = request.headers.get('X-IW-Signature')
    timestamp = request.headers.get('X-IW-Timestamp')
    idempotency_key = request.headers.get('X-IW-Idempotency-Key')

    # Validate required headers
    if not signature or not timestamp:
        return jsonify({
            'error': 'Missing required headers',
            'required': ['X-IW-Signature', 'X-IW-Timestamp']
        }), 400

    # Verify timestamp
    try:
        timestamp_int = int(timestamp)
    except ValueError:
        return jsonify({
            'error': 'Invalid timestamp format'
        }), 400

    if not is_timestamp_valid(timestamp_int):
        return jsonify({
            'error': 'Invalid timestamp',
            'message': 'Timestamp must be within 5 minutes of current time'
        }), 400

    # Verify signature
    if not verify_signature(WEBHOOK_SECRET, timestamp, request.json, signature):
        return jsonify({
            'error': 'Invalid signature',
            'message': 'HMAC-SHA256 signature verification failed'
        }), 401

    # Check idempotency
    if idempotency_key:
        processed = check_idempotency(idempotency_key)
        if processed:
            return jsonify({
                'message': 'Already processed',
                'cached': processed
            }), 200

    # Process webhook
    try:
        result = await process_webhook(connector, request.json)
        
        # Store idempotency key
        if idempotency_key:
            store_idempotency(idempotency_key, result)

        return jsonify({
            'success': True,
            'result': result
        }), 200
    except Exception as error:
        print(f'Webhook processing error: {error}')
        return jsonify({
            'error': 'Processing failed',
            'message': str(error)
        }), 500


@app.route('/health', methods=['GET'])
def health_check():
    """Health check endpoint."""
    return jsonify({'status': 'ok'}), 200


if __name__ == '__main__':
    port = int(os.environ.get('PORT', 3000))
    app.run(host='0.0.0.0', port=port, debug=False)

File: examples/js/webhook-verification.js
/**
 * Webhook Verification Example (Node.js)
 * 
 * Demonstrates HMAC-SHA256 signature verification, timestamp validation,
 * and idempotency checking for IntegrateWise OS webhooks.
 */

const crypto = require('crypto');
const express = require('express');

const app = express();
const WEBHOOK_SECRET = process.env.WEBHOOK_SECRET || 'your-webhook-secret';
const TIMESTAMP_TOLERANCE = 300; // 5 minutes in seconds
const REPLAY_WINDOW = 300; // 5 minutes in seconds

// In-memory idempotency store (use Redis/DB in production)
const idempotencyStore = new Map();

app.use(express.json({
  verify: (req, res, buf) => {
    req.rawBody = buf;
  }
}));

/**
 * Verify timestamp is within tolerance window
 */
function isTimestampValid(timestamp, toleranceSeconds = TIMESTAMP_TOLERANCE) {
  const now = Math.floor(Date.now() / 1000);
  const diff = Math.abs(now - timestamp);
  return diff <= toleranceSeconds;
}

/**
 * Calculate HMAC-SHA256 signature
 */
function calculateSignature(secret, timestamp, body) {
  const payload = `${timestamp}.${JSON.stringify(body)}`;
  const hmac = crypto.createHmac('sha256', secret);
  hmac.update(payload);
  return `sha256=${hmac.digest('hex')}`;
}

/**
 * Verify signature using timing-safe comparison
 */
function verifySignature(secret, timestamp, body, signature) {
  const expected = calculateSignature(secret, timestamp, body);
  return crypto.timingSafeEqual(
    Buffer.from(signature),
    Buffer.from(expected)
  );
}

/**
 * Check if idempotency key was already processed
 */
async function checkIdempotency(idempotencyKey) {
  const stored = idempotencyStore.get(idempotencyKey);
  if (!stored) {
    return null;
  }

  // Check if within replay window
  const now = Date.now();
  const age = now - stored.timestamp;
  if (age > REPLAY_WINDOW * 1000) {
    idempotencyStore.delete(idempotencyKey);
    return null;
  }

  return stored.response;
}

/**
 * Store idempotency key with response
 */
async function storeIdempotency(idempotencyKey, response) {
  idempotencyStore.set(idempotencyKey, {
    timestamp: Date.now(),
    response: response
  });

  // Cleanup old entries (in production, use TTL)
  setTimeout(() => {
    idempotencyStore.delete(idempotencyKey);
  }, REPLAY_WINDOW * 1000);
}

/**
 * Process webhook payload
 */
async function processWebhook(connector, payload) {
  // Simulate processing
  console.log(`Processing webhook from ${connector}:`, payload);
  
  // In real implementation:
  // 1. Parse payload
  // 2. Map to Spine entity
  // 3. Validate constraints
  // 4. Store in Spine
  // 5. Trigger downstream processing
  
  return { success: true, processedAt: new Date().toISOString() };
}

/**
 * Webhook endpoint handler
 */
app.post('/webhooks/:connector', async (req, res) => {
  const { connector } = req.params;
  const signature = req.headers['x-iw-signature'];
  const timestamp = req.headers['x-iw-timestamp'];
  const idempotencyKey = req.headers['x-iw-idempotency-key'];

  // Validate required headers
  if (!signature || !timestamp) {
    return res.status(400).json({
      error: 'Missing required headers',
      required: ['X-IW-Signature', 'X-IW-Timestamp']
    });
  }

  // Verify timestamp
  const timestampInt = parseInt(timestamp, 10);
  if (isNaN(timestampInt) || !isTimestampValid(timestampInt)) {
    return res.status(400).json({
      error: 'Invalid timestamp',
      message: 'Timestamp must be within 5 minutes of current time'
    });
  }

  // Verify signature
  if (!verifySignature(WEBHOOK_SECRET, timestamp, req.body, signature)) {
    return res.status(401).json({
      error: 'Invalid signature',
      message: 'HMAC-SHA256 signature verification failed'
    });
  }

  // Check idempotency
  if (idempotencyKey) {
    const processed = await checkIdempotency(idempotencyKey);
    if (processed) {
      return res.status(200).json({
        message: 'Already processed',
        cached: processed
      });
    }
  }

  // Process webhook
  try {
    const result = await processWebhook(connector, req.body);
    
    // Store idempotency key
    if (idempotencyKey) {
      await storeIdempotency(idempotencyKey, result);
    }

    res.status(200).json({
      success: true,
      result: result
    });
  } catch (error) {
    console.error('Webhook processing error:', error);
    res.status(500).json({
      error: 'Processing failed',
      message: error.message
    });
  }
});

// Health check endpoint
app.get('/health', (req, res) => {
  res.status(200).json({ status: 'ok' });
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Webhook server listening on port ${PORT}`);
});

module.exports = {
  isTimestampValid,
  calculateSignature,
  verifySignature,
  checkIdempotency,
  storeIdempotency
};
